<h3>Load numpy and pandas</h3>

In [2]:
import numpy as np
import pandas as pd

<h3>Load source csv file</h3>

In [3]:
df = pd.read_csv(r'/Users/jackthefob/Downloads/signet.csv')

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x84 in position 8: invalid start byte

<h6>After some online search I solved the data loading problem
source: https://www.edureka.co/community/66416/unicodedecodeerror-codec-decode-position-invalid-start-byte </h6>

In [4]:
df = pd.read_csv(r'/Users/jackthefob/Downloads/signet.csv', encoding='ISO-8859–1')

In [5]:
df.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

         ORDERDATE STATUS  QTR_ID Month-Year  PRODUCTLINE  ...  \
0   2/24/2003 0:00  Shppd       1     2-2003  Motorcycles  ...   
1    5/7/2003 0:00  Shppd       2     5-2003  Motorcycles  ...   
2    7/1/2003 0:00  Shppd       3     7-2003  Motorcycles  ...   
3   8/25/2003 0:00  Shppd       3     8-2003  Motorcycles  ...   
4  10/10/2003 0:00  Shppd       4    10-2003  Motorcycles  ...   

                    ADDRESSLINE1 ADDRESSLINE2           CITY STATE POSTALCODE  \
0        897 Long Airport Avenue          NaN            NYC    NY      10022   
1             59 rue de l'Abbaye          NaN          Reims   NaN      51100   
2  27 rue du Colonel Pierre Avia          NaN          Paris   NaN      75508   
3             78934 Hillside Dr.          NaN       Pasadena    CA      90003   
4                7734 Strong St.          NaN  San Francisco    CA        NaN   

  COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0     USA       NaN              Yu             Kwai    Small  
1  France      EMEA         Henriot             Paul    Small  
2  France      EMEA        Da Cunha           Daniel   Medium  
3     USA       NaN           Young            Julie   Medium  
4     USA       NaN           Brown            Julie   Medium  

[5 rows x 24 columns]

<h3>Use df.(column name).unique() to find typos</h3>

In [6]:
df.STATUS.unique()

array(['Shppd', 'Disputed', 'In Process', 'Cancelled', 'On Hold',
       'Resolved', 'Canceled', 'Shipped'], dtype=object)

<h5>Is it Canceled or Cancelled? Grammarly says canceled = American; cancelled = British
source: https://www.grammarly.com/blog/canceled-vs-cancelled/#:~:text=Canceled%20or%20cancelled%20is%20the,no%20matter%20where%20you%20are.</h5>

In [7]:
df['STATUS'].value_counts()

Shipped       1534
Shppd         1083
Resolved        47
Canceled        45
On Hold         44
In Process      41
Cancelled       15
Disputed        14
Name: STATUS, dtype: int64

<h3>We need to replace 'Shppd' with 'Shipped' and 'Cancelled' with 'Canceled' </h3>

In [8]:
df["STATUS"]= df["STATUS"].replace("Shppd", "Shipped")

In [9]:
df["STATUS"]= df["STATUS"].replace("Cancelled", "Canceled")

In [10]:
df['STATUS'].value_counts()

Shipped       2617
Canceled        60
Resolved        47
On Hold         44
In Process      41
Disputed        14
Name: STATUS, dtype: int64

<h3>No typo was observed in the column PRODUCTLINE</h3>

In [11]:
df['PRODUCTLINE'].value_counts()

Classic Cars        967
Vintage Cars        607
Motorcycles         331
Planes              306
Trucks and Buses    301
Ships               234
Trains               77
Name: PRODUCTLINE, dtype: int64

<h3>No typo was observed in the column CITY</h3>

In [12]:
df['CITY'].value_counts()

Madrid         304
San Rafael     180
NYC            152
Singapore       79
Paris           70
              ... 
Sevilla         15
Los Angeles     14
Munich          14
Burbank         13
Charleroi        8
Name: CITY, Length: 73, dtype: int64

In [13]:
df.CITY.unique()

array(['NYC', 'Reims', 'Paris', 'Pasadena', 'San Francisco', 'Burlingame',
       'Lille', 'Bergen', 'Melbourne', 'Newark', 'Bridgewater', 'Nantes',
       'Cambridge', 'Helsinki', 'Stavern', 'Allentown', 'Salzburg',
       'Chatswood', 'New Bedford', 'Liverpool', 'Madrid', 'Lule',
       'Singapore', 'South Brisbane', 'Philadelphia', 'Lyon', 'Vancouver',
       'Burbank', 'New Haven', 'Minato-ku', 'Torino', 'Boras',
       'Versailles', 'San Rafael', 'Nashua', 'Brickhaven', 'North Sydney',
       'Montreal', 'Osaka', 'White Plains', 'Kobenhavn', 'London',
       'Toulouse', 'Barcelona', 'Los Angeles', 'San Diego', 'Bruxelles',
       'Tsawassen', 'Boston', 'Cowes', 'Oulu', 'San Jose', 'Graz',
       'Makati City', 'Marseille', 'Koln', 'Gensve', 'Reggio Emilia',
       'Frankfurt', 'Espoo', 'Dublin', 'Manchester', 'Aaarhus',
       'Glendale', 'Sevilla', 'Brisbane', 'Strasbourg', 'Las Vegas',
       'Oslo', 'Bergamo', 'Glen Waverly', 'Munich', 'Charleroi'],
      dtype=object)

<h3>No typo was observed in the column STATE</h3>

In [14]:
df['STATE'].value_counts()

CA               416
MA               190
NY               178
NSW               92
Victoria          78
PA                75
CT                61
BC                48
NH                34
Tokyo             32
NV                29
Isle of Wight     26
Quebec            22
NJ                21
Osaka             20
Queensland        15
Name: STATE, dtype: int64

<h3>No typo was observed in the column COUNTRY</h3>

In [15]:
df['COUNTRY'].value_counts()

USA            1004
Spain           342
France          314
Australia       185
UK              144
Italy           113
Finland          92
Norway           85
Singapore        79
Canada           70
Denmark          63
Germany          62
Sweden           57
Austria          55
Japan            52
Belgium          33
Switzerland      31
Philippines      26
Ireland          16
Name: COUNTRY, dtype: int64

<h3>No typo was observed in the column TERRITORY</h3>

In [16]:
df['TERRITORY'].value_counts()

EMEA     1407
APAC      221
Japan     121
Name: TERRITORY, dtype: int64

<h3>No typo was observed in the column DEALSIZE</h3>

In [17]:
df['DEALSIZE'].value_counts()

Medium    1384
Small     1282
Large      157
Name: DEALSIZE, dtype: int64

<h3>Remove the timestamp from ORDERDATE column (keeping mm/dd/YYYY format)</h3>

In [18]:
df['ORDERDATE']= pd.to_datetime(df['ORDERDATE'])

In [19]:
df['ORDERDATE'] = df['ORDERDATE'].dt.strftime('%m/%d/%Y')

In [20]:
df['ORDERDATE'].head()

0    02/24/2003
1    05/07/2003
2    07/01/2003
3    08/25/2003
4    10/10/2003
Name: ORDERDATE, dtype: object

<h3>Add a dollar sign to following columns:</h3>

In [21]:
df['PRICEEACH'] = df['PRICEEACH'].map("${:.2f}".format)

In [22]:
df['PRICEEACH'].head()

0     $95.70
1     $81.35
2     $94.74
3     $83.26
4    $100.00
Name: PRICEEACH, dtype: object

In [23]:
df['SALES'] = df['SALES'].map("${:.2f}".format)

In [24]:
df['SALES'].head()

0    $2871.00
1    $2765.90
2    $3884.34
3    $3746.70
4    $5205.27
Name: SALES, dtype: object

In [25]:
df['MSRP'] = df['MSRP'].map("${:.0f}".format)

In [26]:
df['MSRP'].head()

0    $95
1    $95
2    $95
3    $95
4    $95
Name: MSRP, dtype: object

<h3>Separate the Month and Year into two separate columns</h3>

In [27]:
df['Month'] = pd.DatetimeIndex(df['ORDERDATE']).month

In [28]:
df['Month'].head()

0     2
1     5
2     7
3     8
4    10
Name: Month, dtype: int64

In [29]:
df['Year'] = pd.DatetimeIndex(df['ORDERDATE']).year

In [30]:
df['Year'].head()

0    2003
1    2003
2    2003
3    2003
4    2003
Name: Year, dtype: int64

In [31]:
df['Year'].tail()

2818    2004
2819    2005
2820    2005
2821    2005
2822    2005
Name: Year, dtype: int64

In [32]:
mid = df['Month']
df.drop(labels=['Month'], axis=1, inplace = True)
df.insert(8, 'Month', mid)

In [33]:
mid = df['Year']
df.drop(labels=['Year'], axis=1, inplace = True)
df.insert(9, 'Year', mid)

In [34]:
df.drop(labels=['Month-Year'], axis=1, inplace = True)

<h3>Q1: What is the population average quantity per order?</h3>

In [35]:
df['QUANTITYORDERED'].mean()

35.09280906836698

<h3>Ans: 35.09 per order</h3>

<h3>Q2: What is our top selling product?</h3>

In [37]:
df.groupby('PRODUCTCODE').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'])

QUANTITYORDERED
PRODUCTCODE                 
S18_4933                 714
S24_1046                 724
S18_2248                 743
S24_3969                 745
S18_2581                 746
...                      ...
S12_4473                1024
S700_4002               1029
S18_4600                1031
S24_3856                1052
S18_3232                1774

[109 rows x 1 columns]

<h3>Ans: Our top selling product is S18_3232 (quantity ordered = 1774)</h3>

<h3>Q3: What is the average cancellation rate?</h3>

In [38]:
len(df[df['STATUS'] == 'Canceled'].index)/len(df.index)

0.021253985122210415

<h3>Ans: The cancellation rate is 2.13%</h3>

<h3>Q4: Put together summary metrics and a brief description of your observations for each type of DEAL SIZE</h3>

<h4>This should include, at minimum: sample averages and a list of the top 5 clients for each category

- What was our strongest sales period by volume? By value?

- What is our top selling region by volume? Is this the same as by value?</h4>

In [39]:
dfSmall = df[df['DEALSIZE'] == 'Small']

In [40]:
dfSmall.head()

ORDERNUMBER  QUANTITYORDERED PRICEEACH  ORDERLINENUMBER     SALES  \
0         10107               30    $95.70                2  $2871.00   
1         10121               34    $81.35                5  $2765.90   
6         10180               29    $86.13                9  $2497.77   
8         10201               22    $98.57                2  $2168.54   
11        10237               23   $100.00                7  $2333.12   

     ORDERDATE   STATUS  QTR_ID  Month  Year  ...               ADDRESSLINE1  \
0   02/24/2003  Shipped       1      2  2003  ...    897 Long Airport Avenue   
1   05/07/2003  Shipped       2      5  2003  ...         59 rue de l'Abbaye   
6   11/11/2003  Shipped       4     11  2003  ...    184, chausse de Tournai   
8   12/01/2003  Shipped       4     12  2003  ...  5557 North Pendale Street   
11  04/05/2004  Shipped       2      4  2004  ...          2678 Kingston Rd.   

   ADDRESSLINE2           CITY STATE POSTALCODE COUNTRY TERRITORY  \
0           NaN            NYC    NY      10022     USA       NaN   
1           NaN          Reims   NaN      51100  France      EMEA   
6           NaN          Lille   NaN      59000  France      EMEA   
8           NaN  San Francisco    CA        NaN     USA       NaN   
11    Suite 101            NYC    NY      10022     USA       NaN   

   CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0               Yu             Kwai    Small  
1          Henriot             Paul    Small  
6            Rance          Martine    Small  
8           Murphy            Julie    Small  
11           Frick          Michael    Small  

[5 rows x 25 columns]

In [41]:
dfMedium = df[df['DEALSIZE'] == 'Medium']

In [42]:
dfMedium.head()

ORDERNUMBER  QUANTITYORDERED PRICEEACH  ORDERLINENUMBER     SALES  \
2        10134               41    $94.74                2  $3884.34   
3        10145               45    $83.26                6  $3746.70   
4        10159               49   $100.00               14  $5205.27   
5        10168               36    $96.66                1  $3479.76   
7        10188               48   $100.00                1  $5512.32   

    ORDERDATE   STATUS  QTR_ID  Month  Year  ...  \
2  07/01/2003  Shipped       3      7  2003  ...   
3  08/25/2003  Shipped       3      8  2003  ...   
4  10/10/2003  Shipped       4     10  2003  ...   
5  10/28/2003  Shipped       4     10  2003  ...   
7  11/18/2003  Shipped       4     11  2003  ...   

                    ADDRESSLINE1 ADDRESSLINE2           CITY STATE POSTALCODE  \
2  27 rue du Colonel Pierre Avia          NaN          Paris   NaN      75508   
3             78934 Hillside Dr.          NaN       Pasadena    CA      90003   
4                7734 Strong St.          NaN  San Francisco    CA        NaN   
5              9408 Furth Circle          NaN     Burlingame    CA      94217   
7    Drammen 121, PR 744 Sentrum          NaN         Bergen   NaN     N 5804   

  COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
2  France      EMEA        Da Cunha           Daniel   Medium  
3     USA       NaN           Young            Julie   Medium  
4     USA       NaN           Brown            Julie   Medium  
5     USA       NaN          Hirano             Juri   Medium  
7  Norway      EMEA          Oeztan           Veysel   Medium  

[5 rows x 25 columns]

In [43]:
dfLarge = df[df['DEALSIZE'] == 'Large']

In [44]:
dfLarge.head()

ORDERNUMBER  QUANTITYORDERED PRICEEACH  ORDERLINENUMBER     SALES  \
20        10341               41   $100.00                9  $7737.93   
25        10417               66   $100.00                2  $7516.08   
27        10112               29   $100.00                1  $7209.11   
28        10126               38   $100.00               11  $7329.06   
29        10140               37   $100.00               11  $7374.10   

     ORDERDATE    STATUS  QTR_ID  Month  Year  ...        ADDRESSLINE1  \
20  11/24/2004   Shipped       4     11  2004  ...         Geislweg 14   
25  05/13/2005  Disputed       2      5  2005  ...  C/ Moralzarzal, 86   
27  03/24/2003   Shipped       1      3  2003  ...     Berguvsv„gen  8   
28  05/28/2003   Shipped       2      5  2003  ...      C/ Araquil, 67   
29  07/24/2003   Shipped       3      7  2003  ...   9408 Furth Circle   

   ADDRESSLINE2        CITY STATE POSTALCODE  COUNTRY TERRITORY  \
20          NaN    Salzburg   NaN       5020  Austria      EMEA   
25          NaN      Madrid   NaN      28034    Spain      EMEA   
27          NaN        Lule   NaN   S-958 22   Sweden      EMEA   
28          NaN      Madrid   NaN      28023    Spain      EMEA   
29          NaN  Burlingame    CA      94217      USA       NaN   

   CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
20           Pipps            Georg    Large  
25          Freyre            Diego    Large  
27        Berglund        Christina    Large  
28          Sommer           Mart¡n    Large  
29          Hirano             Juri    Large  

[5 rows x 25 columns]

<h4>Sample average by quantity ordered</h4>

In [45]:
dfSmall['QUANTITYORDERED'].mean()

30.525741029641186

In [46]:
dfMedium['QUANTITYORDERED'].mean()

37.94725433526011

In [47]:
dfLarge['QUANTITYORDERED'].mean()

47.22292993630573

<h4>Sample average by Sales</h4>

In [48]:
df['SALES'] = df['SALES'].str.replace('$','').astype('float')

In [49]:
df[df['DEALSIZE'] == 'Small']['SALES'].mean()

2061.68280031201

In [50]:
df[df['DEALSIZE'] == 'Medium']['SALES'].mean()

4398.433699421969

In [51]:
df[df['DEALSIZE'] == 'Large']['SALES'].mean()

8293.753248407647

<h4>Identify top 5 clients for each category</h4>

In [52]:
df[df['DEALSIZE'] == 'Small'].groupby('CUSTOMERNAME').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
CUSTOMERNAME                                 
Euro Shopping Channel                    3685
Mini Gifts Distributors Ltd.             2443
Rovelli Gifts                             877
AV Stores, Co.                            834
Australian Collectors, Co.                806
...                                       ...
Atelier graphique                         125
Auto-Moto Classics Inc.                   120
Super Scale Inc.                          112
Royale Belge                               75
Boards & Toys Co.                          67

[92 rows x 1 columns]

In [53]:
df[df['DEALSIZE'] == 'Small'].groupby('CUSTOMERNAME').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
CUSTOMERNAME                           
Euro Shopping Channel         237852.28
Mini Gifts Distributors Ltd.  153135.56
Australian Collectors, Co.     57325.62
Rovelli Gifts                  56972.88
Land of Toys Inc.              53253.52
...                                 ...
Online Mini Collectables        8019.81
Super Scale Inc.                7940.51
Atelier graphique               6855.90
Royale Belge                    5660.70
Boards & Toys Co.               5142.15

[92 rows x 1 columns]

In [54]:
df[df['DEALSIZE'] == 'Medium'].groupby('CUSTOMERNAME').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
CUSTOMERNAME                                       
Euro Shopping Channel                          5064
Mini Gifts Distributors Ltd.                   3432
La Rochelle Gifts                              1116
Muscle Machine Inc                             1062
Australian Collectors, Co.                     1025
...                                             ...
Auto-Moto Classics Inc.                         167
Microscale Inc.                                 164
Bavarian Collectables Imports, Co.              159
Atelier graphique                               145
Boards & Toys Co.                                35

[92 rows x 1 columns]

In [56]:
df[df['DEALSIZE'] == 'Medium'].groupby('CUSTOMERNAME').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
CUSTOMERNAME                                 
Euro Shopping Channel               580216.42
Mini Gifts Distributors Ltd.        408562.74
Muscle Machine Inc                  128034.90
Australian Collectors, Co.          124630.90
La Rochelle Gifts                   122367.40
...                                       ...
Bavarian Collectables Imports, Co.   17522.44
Auto-Moto Classics Inc.              17384.03
Atelier graphique                    17324.06
Microscale Inc.                      14677.60
Boards & Toys Co.                     3987.20

[92 rows x 1 columns]

In [57]:
df[df['DEALSIZE'] == 'Large'].groupby('CUSTOMERNAME').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
CUSTOMERNAME                                 
Euro Shopping Channel                     578
Mini Gifts Distributors Ltd.              491
The Sharp Gifts Warehouse                 378
Mini Caravy                               269
Danish Wholesale Imports                  265
...                                       ...
Scandinavian Gift Ideas                    37
Marta's Replicas Co.                       37
Technics Stores Inc.                       37
Australian Gift Network, Co                34
Volvo Model Replicas, Co                   29

[70 rows x 1 columns]

In [58]:
df[df['DEALSIZE'] == 'Large'].groupby('CUSTOMERNAME').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
CUSTOMERNAME                          
Euro Shopping Channel         94225.41
Mini Gifts Distributors Ltd.  93159.76
The Sharp Gifts Warehouse     53442.10
Danish Wholesale Imports      41911.23
Muscle Machine Inc            41328.10
...                                ...
Volvo Model Replicas, Co       7209.11
Land of Toys Inc.              7140.76
Scandinavian Gift Ideas        7136.19
Diecast Collectables           7060.24
Handji Gifts& Co               7020.48

[70 rows x 1 columns]

<h3>What was our strongest sales period by volume? By value?</h3>

In [61]:
df.groupby('Month').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
Month                 
11               20857
10               10998
5                 8992
1                 7997
2                 7903
3                 7585
4                 6704
8                 6538
12                6293
9                 5681
7                 4899
6                 4620

In [62]:
df.groupby('Month').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
Month            
11     2118885.67
10     1121215.22
5       923972.56
2       810441.90
1       785874.44
3       754501.39
4       669390.96
8       659310.57
12      634679.12
9       584724.27
7       514875.97
6       454756.78

In [63]:
df.groupby('Year').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
Year                 
2004            46824
2003            34612
2005            17631

In [64]:
df.groupby('Year').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
Year            
2004  4724162.60
2003  3516979.54
2005  1791486.71

In [65]:
df.groupby(['Month','Year']).agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
Month Year                 
11    2004            10678
      2003            10179
10    2003             5515
      2004             5483
8     2004             4564
5     2005             4357
3     2005             3852
12    2004             3804
1     2005             3395
2     2005             3393
1     2004             3245
7     2004             3174
9     2004             3171
2     2004             3061
6     2004             2971
4     2005             2634
5     2004             2618
9     2003             2510
12    2003             2489
4     2004             2077
5     2003             2017
4     2003             1993
3     2004             1978
8     2003             1974
3     2003             1755
7     2003             1725
6     2003             1649
2     2003             1449
1     2003             1357

In [66]:
df.groupby(['Month','Year']).agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
Month Year            
11    2004  1089048.01
      2003  1029837.66
10    2003   568290.97
      2004   552924.25
8     2004   461501.27
5     2005   457861.06
3     2005   374262.76
12    2004   372802.66
2     2005   358186.18
1     2005   339543.42
7     2004   327144.09
9     2004   320750.91
1     2004   316577.42
2     2004   311419.53
6     2004   286674.22
5     2004   273438.39
9     2003   263973.36
12    2003   261876.46
4     2005   261633.29
      2004   206148.12
3     2004   205733.73
4     2003   201609.55
8     2003   197809.30
5     2003   192673.11
7     2003   187731.88
3     2003   174504.90
6     2003   168082.56
2     2003   140836.19
1     2003   129753.60

<h3>Ans:
1. By month: November (both volume and value)
2. By year: 2004 (both volume and value)
3. By month AND year: November 2004 (both volume and value)</h3>

<h3>What is our top selling region by volume? Is this the same as by value?</h3>

In [67]:
df.groupby('CITY').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
CITY                        
Madrid                 10958
San Rafael              6366
NYC                     5294
Singapore               2760
Paris                   2521
...                      ...
Burbank                  511
Los Angeles              500
Dublin                   490
Munich                   401
Charleroi                278

[73 rows x 1 columns]

In [68]:
df.groupby('CITY').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
CITY                   
Madrid       1082551.44
San Rafael    654858.06
NYC           560787.77
Singapore     288488.41
Paris         268944.68
...                 ...
Brisbane       50218.51
Los Angeles    48048.46
Burbank        46084.64
Munich         34993.92
Charleroi      33440.10

[73 rows x 1 columns]

<h4>The answer is Madrid (both volume and value) if we group by city</h4>

In [69]:
df.groupby('POSTALCODE').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
POSTALCODE                 
28034                  9327
97562                  7269
10022                  5294
94217                  3349
44000                  2102
...                     ...
2                       490
80686                   401
WA1 1DP                 357
B-6000                  278
92561                   102

[73 rows x 1 columns]

In [70]:
df[df['POSTALCODE'] == '28034']

ORDERNUMBER  QUANTITYORDERED PRICEEACH  ORDERLINENUMBER     SALES  \
25          10417               66   $100.00                2   7516.08   
53          10424               50   $100.00                6  12001.00   
79          10417               45   $100.00                5   5887.35   
105         10417               56   $100.00                4   9218.16   
126         10350               26    $75.47                5   1962.22   
...           ...              ...       ...              ...       ...   
2782        10262               21    $57.11               12   1199.31   
2793        10386               50    $87.15               16   4357.50   
2801        10153               50    $60.06                4   3003.00   
2818        10350               20   $100.00               15   2244.40   
2820        10386               43   $100.00                4   5417.57   

       ORDERDATE      STATUS  QTR_ID  Month  Year  ...        ADDRESSLINE1  \
25    05/13/2005    Disputed       2      5  2005  ...  C/ Moralzarzal, 86   
53    05/31/2005  In Process       2      5  2005  ...  C/ Moralzarzal, 86   
79    05/13/2005    Disputed       2      5  2005  ...  C/ Moralzarzal, 86   
105   05/13/2005    Disputed       2      5  2005  ...  C/ Moralzarzal, 86   
126   12/02/2004     Shipped       4     12  2004  ...  C/ Moralzarzal, 86   
...          ...         ...     ...    ...   ...  ...                 ...   
2782  06/24/2004    Canceled       2      6  2004  ...  C/ Moralzarzal, 86   
2793  03/01/2005    Resolved       1      3  2005  ...  C/ Moralzarzal, 86   
2801  09/28/2003     Shipped       3      9  2003  ...  C/ Moralzarzal, 86   
2818  12/02/2004     Shipped       4     12  2004  ...  C/ Moralzarzal, 86   
2820  03/01/2005    Resolved       1      3  2005  ...  C/ Moralzarzal, 86   

     ADDRESSLINE2    CITY STATE POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME  \
25            NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
53            NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
79            NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
105           NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
126           NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
...           ...     ...   ...        ...     ...       ...             ...   
2782          NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
2793          NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
2801          NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
2818          NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   
2820          NaN  Madrid   NaN      28034   Spain      EMEA          Freyre   

     CONTACTFIRSTNAME DEALSIZE  
25              Diego    Large  
53              Diego    Large  
79              Diego   Medium  
105             Diego    Large  
126             Diego    Small  
...               ...      ...  
2782            Diego    Small  
2793            Diego   Medium  
2801            Diego   Medium  
2818            Diego    Small  
2820            Diego   Medium  

[259 rows x 25 columns]

In [71]:
df.groupby('POSTALCODE').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
POSTALCODE           
28034       912294.11
97562       756752.85
10022       560787.77
94217       331011.85
50553       207874.86
...               ...
8010         52263.90
WA1 1DP      36019.04
80686        34993.92
B-6000       33440.10
92561         9129.35

[73 rows x 1 columns]

<h4>The answer is still Madrid (both volume and value) if we sort by POSTALCODE</h4>

In [72]:
df.groupby('COUNTRY').agg({'QUANTITYORDERED':'sum'}).sort_values(by=['QUANTITYORDERED'], ascending=False)

QUANTITYORDERED
COUNTRY                     
USA                    35659
Spain                  12429
France                 11090
Australia               6246
UK                      5013
Italy                   3773
Finland                 3192
Norway                  2842
Singapore               2760
Canada                  2293
Denmark                 2197
Germany                 2148
Sweden                  2006
Austria                 1974
Japan                   1842
Switzerland             1078
Belgium                 1074
Philippines              961
Ireland                  490

In [73]:
df.groupby('COUNTRY').agg({'SALES':'sum'}).sort_values(by=['SALES'], ascending=False)

SALES
COUNTRY                
USA          3627982.83
Spain        1215686.92
France       1110916.52
Australia     630623.10
UK            478880.46
Italy         374674.31
Finland       329581.91
Norway        307463.70
Singapore     288488.41
Denmark       245637.15
Canada        224078.56
Germany       220472.09
Sweden        210014.21
Austria       202062.53
Japan         188167.81
Switzerland   117713.56
Belgium       108412.62
Philippines    94015.73
Ireland        57756.43

<h4>The answer is USA (both volume and value) if we sort by COUNTRY</h4>

<h3>We need to add the dollar sign back to the values on SALES column before we save the cleaned dataframe</h3>

In [75]:
df['SALES'].head()

0    2871.00
1    2765.90
2    3884.34
3    3746.70
4    5205.27
Name: SALES, dtype: float64

In [76]:
df['SALES'] = df['SALES'].map("${:.2f}".format)

In [77]:
df['SALES'].head()

0    $2871.00
1    $2765.90
2    $3884.34
3    $3746.70
4    $5205.27
Name: SALES, dtype: object

In [78]:
df.to_csv('YuHonLinCleanedDataframe.csv')